### PR TITLE
D3 QA, OOP & FP intros, new Vis Des & Resp Des challenges

### PR DESCRIPTION
**Pull request by HKuz**
#### Pre-Submission Checklist

- [x] Your pull request targets the `feature/curriculum-expansion` branch of FreeCodeCamp (not `staging`)
- [x] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [x] You have only one commit (if not, [squash](http://forum.freecodecamp.com/t/how-to-squash-multiple-commits-into-one-with-git/13231) them into one commit).
- [ ] All new and existing tests pass the command `npm test`. Use `git commit --amend` to amend any fixes. [Tests don't work in general on this branch]

#### Type of Change

- [x] Small bug fix (non-breaking change which fixes an issue)

#### Checklist:

- [x] Tested changes locally.

#### Description

- D3 section QA (text changed for consistency and clarity), added and checked assert tests for all challenges. Also adds the CDN script source in the background for all challenges. (Note - running this section locally was a bit buggy, a page reload was required every few challenges to get the output to show in the phone and the tests to pass. Not clear if this is just a local issue.)
- Intros added/edited for the Functional Programming and Object Oriented Programming sections
- New challenge added in Visual design, along with some text changes to switch to 2nd person
- Challenge edit in Responsive Web design, along with some text changes to switch to 2nd person

### CODE REVIEW

File: seed/challenges/02-javascript-algorithms-and-data-structures/functional-programming.json [hunk 1]
```diff
@@ -40,9 +40,9 @@
         "Functional programming is a style of programming where solutions are simple, isolated functions, without any side effects outside of the function scope.",
         "<code>INPUT -> PROCESS -> OUTPUT</code>",
         "Functional programming is about:",
-        "1) Isolated functions - there is no dependence on external state, which includes global variables that are subject to change",
+        "1) Isolated functions - there is no dependence on the state of the program, which includes global variables that are subject to change",
         "2) Pure functions - the same input always gives the same output",
-        "3) Functions have no side effects - they do not mutate (change) external states",
+        "3) Functions have no side effects - they do not mutate, or change, the state of the program outside the function",
```
Comment (terakilobyte) [inline seed/challenges/02-javascript-algorithms-and-data-structures/functional-programming.json]: Sadly, this is a very academic statement and in no way resembles the real world. Functions aim to have no side effects, and most should not. Some will necessarily have to, and they should be carefully controlled. Without some kind of side effects we can't do anything actually useful to the user in the end.

File: seed/challenges/02-javascript-algorithms-and-data-structures/functional-programming.json [hunk 2]
```diff
@@ -40,9 +40,9 @@
...
         "<h4>Instructions</h4>",
         "The members of Free Code Camp happen to love tea.",
         "In the code editor, the <code>prepareTea</code> and <code>getTea</code> functions are already defined for you. Call the <code>getTea</code> function to get 40 cups of tea for the team, and store them in the <code>tea4TeamFCC</code> variable."
```
Comment (QuincyLarson) [inline seed/challenges/02-javascript-algorithms-and-data-structures/functional-programming.json]: @terakilobyte your suggestions are appreciated. Can you commit to going through this section and completely overhauling the challenges yourself?

**Comment by HKuz** [comment]
@terakilobyte - updated PR includes some tweaks to the FP intro and the challenge bullet to address your 1st comment about real world vs academic realities. I saw your post on the issue thread about your other point, and can respond there. 👍

**Comment by raisedadead** [comment]
@terakilobyte bump for your review.

**Comment by QuincyLarson** [comment]
I'm merging this since the same CI issues are on the curriculum expansion branch as well. Then we'll resolve those there after we've merged in staging.

**Pull request merged**